src/boundaries/stagger_and_step.jl

```diff
@@ -0,0 +1,250 @@
+from a region defined by the user (see the argument `isinside`).
+
+Given the dynamical mapping `F`, if the iteration `x_{n+1} = 
+F(x_n)` respects the condition `T(x_{n+1}) > Tm` we accept 
+this next point, this is the _step_ part of the method. If 
+not, the method search randomly the next point in a 
+neighborhood following a given probability distribution, this 
+is the _stagger_ part. This part sometimes fails to find a new 
+candidate and a new starting point of the trajectory is chosen 
+within the defined region. See the keyword argument 
+`stagger_mode` for the different available methods.   
+
+The method produces a pseudo-trajectory of `N` points δ-close
+to the stable manifold of the chaotic saddle. 
+
+"""
+function stagger_and_step!(ds::DynamicalSystem, x0, N::Int, isinside::Function; δ = 1e-10, Tm  = 30, 
+    γ = 1.1, max_steps = Int(1e5), max_escape_time = 10000, stagger_mode = :exp, δ₀ = 1., 
+    show_progress = true,  rng = MersenneTwister())
+
+    progress = ProgressMeter.Progress(
+        N; desc = "Saddle estimation: ", dt = 1.0
+    )
+    xi = stagger_trajectory!(ds, x0, Tm, isinside; δ₀, stagger_mode = :unif, 
+                            max_steps, γ, max_escape_time, rng) 
+    if isnothing(xi)
+        error("Cannot find a stagger trajectory. Choose a different starting point or 
+                search radius δ₀.")
+    end
+
+    v = Vector{typeof(current_state(ds))}(undef, N)
+    v[1] = xi
+    for n in 1:N
+        show_progress && ProgressMeter.update!(progress, n)
+        if escape_time!(ds, xi, isinside; max_escape_time) > Tm
+            reinit!(ds, xi; t0 = 0)
+        else
+            xp, Tp = stagger!(ds, xi, δ, Tm, isinside; stagger_mode, max_steps, γ, max_escape_time, rng)
+            # The stagger step may fail. We reinitiate the algorithm from a new initial condition.
+            if Tp < 0
+                xp = stagger_trajectory!(ds, x0, Tm, isinside; δ₀, stagger_mode = :exp,max_steps, γ, max_escape_time, rng) 
+                if isnothing(xp)
+                    error("Cannot find a stagger trajectory. Choose a different starting 
+                          point or search radius δ₀.")
+                end
+                δ = 0.1
+            end
+            reinit!(ds, xp; t0 = 0)
+        end 
+        step!(ds)
+        xi = copy(current_state(ds))
+        v[n] = xi
+    end
+    return v
+end
+
+    
+function escape_time!(ds, x0, isinside; max_escape_time = 10000) 
+    x = copy(x0) 
+    set_state!(ds,x)
+    reinit!(ds, x; t0 = 0)
+    k = 1; 
+    while isinside(x) 
+        if k > max_escape_time 
+            error("The trajectory did not escape for ", k, " steps, you probably have \
+an attractor in the defined region. Last point evaluated: ", x)
+        end
+        step!(ds)
+        x = current_state(ds)
+        k += 1
+    end
+    return current_time(ds)
+end
+
+function rand_u(δ, n; stagger_mode = :exp, rng::AbstractRNG)
+    if stagger_mode == :exp 
```

Comment:
In this internal function you should change `stagger_mode` and `rng` into positional arguments. I think the compiler can do constant propagation more robustly and remove the `if` clause this way.